src/types/mod.rs

```diff
@@ -4,12 +4,31 @@ use anyhow::{Context, Result};
 use bb8_redis::redis;
 use serde::{Deserialize, Serialize};
 use std::io::Write;
+use std::path::PathBuf;
 use std::time::{Duration, SystemTime};
 
 #[derive(Clone, Debug)]
 pub enum TransitMessage {
     Request(Message),
     Reply(Message),
+    Upload(Message),
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct UploadPayload {
+    pub path: PathBuf,
+    pub cmd: String,
+
+    #[serde(default)]
+    #[serde(skip_deserializing)]
+    pub timestamp: u64,
+
+    #[serde(default)]
+    #[serde(skip_deserializing)]
+    pub source: u32,
+
+    #[serde(skip_deserializing)]
+    pub signature: Option<String>,
```

Comment:
I am sorry if this is a lot of work but it will make things much cleaner because now it's really hard to follow the cycle of signing/verification since fields are available in all context while they only be visible in a subset of context.

In other words. path and cmd all available in client, local rmb, remote rmb, server. while signature source, and timestamp are only available in local rmb, and remote rmb. Hence i can't just have all fields everywhere.

If the client can't set the signature or the timestamp, then he shouldn't be able to even set them

Reply:
To make it easy to use the Challengeable interface to calculate the signature you can add this 

```rust
impl<T> Challengeable for &[T]
where
    T: std::fmt::Display,
{
    fn challenge(&self) -> Result<md5::Digest> {
        let mut hash = md5::Context::new();
        for v in self.iter() {
            write!(hash, "{}", v)?;
        }

        Ok(hash.compute())
    }
}
```

then you can simply do 

```rust
let fields = vec![ put all fields here (as strings) ];
sign(&fields);
```